### PR TITLE
[PodSecurity] Metrics improvements

### LINKED ISSUE #103207: [PodSecurity] Implement and test metrics
Tasks:
  - [ ] for both pod/pod-controller admission
  - [ ] record allow/deny/exempt/error cases correctly
  - [ ] record enforce/warn/audit cases correctly

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

1. Fix up current metrics:
    - Register metrics on the global handler
    - Expose a `/metrics` endpoint on the webhook
2. Make metrics conform to the spec defined in https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md#monitoring:
    - `operation` -> `request_operation` for the label name
    - resource should be either `pod` or `controller` (see comment on below)
    - operation should be lower case
3. Add test cases
4. Add the error & exemption metrics

#### Which issue(s) this PR fixes:

Fixes #103207

#### Special notes for your reviewer:

**Question 1:** Do we want to consider the short-circuit case an evaluation, for monitoring purposes? doing so adds a second alloc in the common path, and doubles the evaluation time. See benchmarks below.

**Question 2:** I was corrected in the docs PR (https://github.com/kubernetes/website/pull/29124#pullrequestreview-715714595, @sftim ) that we should refer to the "pod template resources" as workload resources. Do we want to update the resource label on the metric to match (`controller` -> `workload`)?

Benchmarks:
```
# Original (without this PR):

BenchmarkVerifyPod/enforce-implicit_pod-16         	 4421935	       266.8 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-implicit_deployment-16  	 5054050	       267.9 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-privileged_pod-16       	 5110650	       276.4 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-privileged_deployment-16         	 4437080	       280.7 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-baseline_pod-16                  	  315181	      4174 ns/op	    3504 B/op	      19 allocs/op
BenchmarkVerifyPod/enforce-baseline_deployment-16           	 3700531	       303.5 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-restricted_pod-16                	  292236	      4702 ns/op	    3504 B/op	      19 allocs/op
BenchmarkVerifyPod/enforce-restricted_deployment-16         	 4413544	       299.9 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/warn-baseline_pod-16                     	  274810	      4191 ns/op	    3504 B/op	      19 allocs/op
BenchmarkVerifyPod/warn-baseline_deployment-16              	  304512	      3667 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/warn-restricted_pod-16                   	  283743	      4435 ns/op	    3504 B/op	      19 allocs/op
BenchmarkVerifyPod/warn-restricted_deployment-16            	  263958	      3974 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_pod-16       	  187177	      6975 ns/op	    6000 B/op	      29 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_deployment-16         	  256402	      5064 ns/op	    4800 B/op	      24 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_pod-16             	  204177	      5673 ns/op	    4752 B/op	      24 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_deployment-16      	  226470	      5445 ns/op	    4800 B/op	      24 allocs/op

## With this PR

BenchmarkVerifyPod/enforce-implicit_pod-16         	 1946442	       588.9 ns/op	     224 B/op	       2 allocs/op
BenchmarkVerifyPod/enforce-implicit_deployment-16  	 4798802	       256.9 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-privileged_pod-16       	 1862803	       626.7 ns/op	     224 B/op	       2 allocs/op
BenchmarkVerifyPod/enforce-privileged_deployment-16         	 4398561	       283.9 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-baseline_pod-16                  	  301028	      4009 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/enforce-baseline_deployment-16           	 4139908	       284.3 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-restricted_pod-16                	  297513	      4582 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/enforce-restricted_deployment-16         	 3554356	       309.7 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/warn-baseline_pod-16                     	  295238	      4026 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/warn-baseline_deployment-16              	  294944	      3670 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/warn-restricted_pod-16                   	  298869	      4262 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/warn-restricted_deployment-16            	  333310	      3916 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_pod-16       	  166224	      6843 ns/op	    5976 B/op	      28 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_deployment-16         	  257518	      4938 ns/op	    4800 B/op	      24 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_pod-16             	  219632	      5800 ns/op	    4728 B/op	      23 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_deployment-16      	  231926	      5616 ns/op	    4800 B/op	      24 allocs/op

## With this PR, but dropping the short-circuit eval count:

BenchmarkVerifyPod/enforce-implicit_pod-16         	 4651302	       247.4 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-implicit_deployment-16  	 4379590	       265.1 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-privileged_pod-16       	 4283446	       273.9 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-privileged_deployment-16         	 4164979	       282.0 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-baseline_pod-16                  	  307122	      3885 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/enforce-baseline_deployment-16           	 4469337	       289.3 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/enforce-restricted_pod-16                	  300816	      4306 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/enforce-restricted_deployment-16         	 4040166	       283.6 ns/op	     112 B/op	       1 allocs/op
BenchmarkVerifyPod/warn-baseline_pod-16                     	  307626	      3977 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/warn-baseline_deployment-16              	  293635	      3664 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/warn-restricted_pod-16                   	  308919	      4307 ns/op	    3480 B/op	      18 allocs/op
BenchmarkVerifyPod/warn-restricted_deployment-16            	  317541	      3943 ns/op	    3552 B/op	      19 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_pod-16       	  181440	      6677 ns/op	    5976 B/op	      28 allocs/op
BenchmarkVerifyPod/enforce-warn-audit-baseline_deployment-16         	  233512	      5131 ns/op	    4800 B/op	      24 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_pod-16             	  228664	      5582 ns/op	    4728 B/op	      23 allocs/op
BenchmarkVerifyPod/warn-baseline-audit-restricted_deployment-16      	  241128	      5634 ns/op	    4800 B/op	      24 allocs/op

goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/plugin/pkg/admission/security/podsecurity
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
```

#### Does this PR introduce a user-facing change?
```release-note
Add Pod Security admission metrics: pod_security_evaluations_total, pod_security_exemptions_total, pod_security_errors_total
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md#monitoring
```

/sig auth
/milestone v1.23
/assign @liggitt 
/cc @jyz0309

### CODE REVIEW

File: staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go [hunk 1]
```diff
@@ -272,11 +279,14 @@ func Setup(c *Config) (*Server, error) {
 	if err != nil {
 		return nil, fmt.Errorf("could not create PodSecurityRegistry: %w", err)
 	}
+	metrics := metrics.NewPrometheusRecorder(api.GetAPIVersion())
```
[the remark this answers is not in the document]
Comment (tallclair) [reply staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go:282]: Done (just grabbed the first run of numbers)

**Comment by tallclair** [comment]
/hold
Needs squash before merging.

File: staging/src/k8s.io/pod-security-admission/metrics/metrics.go [hunk 2]
```diff
@@ -29,55 +37,245 @@ const (
...
+	r.registerOnce.Do(func() {
+		registerFunc(r.evaluationsCounter)
+		registerFunc(r.exemptionsCounter)
+		registerFunc(r.errorsCounter)
+
+		r.cacheCommonMetrics()
+	})
+}
+
+func (r *PrometheusRecorder) cacheCommonMetrics() {
```
[the remark this answers is not in the document]
Comment (tallclair) [reply staging/src/k8s.io/pod-security-admission/metrics/metrics.go]: Done.

File: staging/src/k8s.io/pod-security-admission/metrics/metrics.go [hunk 3]
```diff
@@ -29,55 +37,245 @@ const (
...
+		{"create", "controller", ""},
+		{"update", "controller", ""},
+	}
+	r.cachedExemptionsCounters = make(map[exemptionsLabels]metrics.CounterMetric, len(cachedExemptionsLabels))
+	for _, l := range cachedExemptionsLabels {
+		r.cachedExemptionsCounters[l] = r.exemptionsCounter.WithLabelValues(l.labels()...)
+	}
+}
+
+func (r *PrometheusRecorder) Reset() {
```
[the remark this answers is not in the document]
Comment (tallclair) [reply staging/src/k8s.io/pod-security-admission/metrics/metrics.go:83]: Done. I agree that it's starting to feel fragile. I think I still prefer this to the alternatives though.

Let me know if you have any ideas for how to deduplicate some of the code (might just need to wait for go 1.18 generics...)

**Comment by tallclair** [comment]
This is ready for another review.

File: staging/src/k8s.io/pod-security-admission/api/helpers.go [hunk 4]
```diff
@@ -76,7 +77,11 @@ func GetAPIVersion() Version {
 	if err != nil {
 		return v
 	}
-	minor, err := strconv.Atoi(apiVersion.Minor)
+	// split the "normal" + and - for semver stuff to get the leading minor number
+	minorString := strings.FieldsFunc(apiVersion.Minor, func(r rune) bool {
+		return !unicode.IsNumber(r)
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/api/helpers.go:82]: https://pkg.go.dev/unicode#IsDigit instead?

**Review by liggitt — COMMENTED** [review]
needs update-vendor.sh run

**Comment by liggitt** [comment]
lgtm otherwise

**Comment by tallclair** [comment]
Done. Also squashed commits & rebased.

/hold cancel

**Comment by tallclair** [comment]
Rebased. Also had to update the test to check specifically for `audit-violations` in the annotations to construct the metrics expectations.

**Pull request merged**